### DISCUSSION TITLE
How do I deserialise an RA-TLS certificate's x509 extension?

### DISCUSSION THREAD

**Discussion by fwoodruff** [gramineproject/gramine#1809] [Q&A]
I need to validate an RA-TLS certificate's x509v3 extension in Typescript.

I need to know what message within the certificate is signed.
I need to parse out a signature (r,s) from the x509 extension.
Finally I need to know what the public key of this signature needs comparison against.

Fundamentally this is a deserialisation exercise! Not complex, just unguessable.

Pointing to various deserialisation steps in the RA-TLS code would be one possible useful answer here.

Are there any other forums where I can get specific RA-TLS support or am I in the right place here?

**Comment by dimakuv** [comment]
First, make sure to read this: https://gramine.readthedocs.io/en/stable/attestation.html#mid-level-ra-tls-interface

Then look at the RA-TLS code:
- https://github.com/gramineproject/gramine/blob/e3d3aa39c11054dfa3ce1ad2ce715ec24d730f1a/tools/sgx/ra-tls/ra_tls_verify_common.c#L122
- https://github.com/gramineproject/gramine/blob/e3d3aa39c11054dfa3ce1ad2ce715ec24d730f1a/tools/sgx/ra-tls/ra_tls_verify_common.c#L222
- https://github.com/gramineproject/gramine/blob/e3d3aa39c11054dfa3ce1ad2ce715ec24d730f1a/tools/sgx/ra-tls/ra_tls_verify_dcap.c#L168 -- this is the call to the external (to Gramine) function from Intel SGX PSW's library `libsgx-dcap-quote-verify`

The currently used OID:
- https://github.com/gramineproject/gramine/blob/e3d3aa39c11054dfa3ce1ad2ce715ec24d730f1a/tools/sgx/ra-tls/ra_tls_common.h#L24-L26

> I need to know what message within the certificate is signed.

I don't know what is meant by this. Inside the certificate, we embed the SGX quote as a special-numbered OID object.

> I need to parse out a signature (r,s) from the x509 extension.

Which signature? Of the SGX quote? You'll need to look at the Intel SGX quote format (the documentation and formats can be found on the internet).

> Finally I need to know what the public key of this signature needs comparison against.

The SGX Quote is signed by the special Quoting Enclave, and it embeds the certificate chain that is rooted at the Intel PCS web service. You can find more documentation on the internet; it's not part of Gramine but rather part of the Intel SGX PSW.

---

I'm unsure what you're trying to achieve. All these "how to verify the SGX quote" is supposed to be hidden by the Intel SGX PSW infrastructure. For example, if you run on MS Azure SGX-enabled VMs, they already have this preinstalled, and you just need to link against `libsgx-dcap-quote-verify` and call a function there, supplying the SGX quote (extracted from the received RA-TLS X.509 certificate). This is what RA-TLS implementation in Gramine performs.
